### ISSUE TITLE
make i0 a compile error

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#1593]
There was a rejected proposal to remove `u0`: #1530 

However, I believe it does make sense to remove `i0`. While the range of unsigned integers is 0 to (pow(2, n) − 1), with n = 0, this gives the range [0, 0]. OK. But the range of signed integers is −(pow(2,n−1)) to (pow(2,n−1) − 1), with n = 0, this makes us do pow(2, -1) and that doesn't make sense.

`i0` would still be an identifier that maps to a signed 0-bit integer. The compile error would happen if you referenced `i0` directly or if you used `@IntType` to construct a signed 0 bit integer.

**Comment by Hejsil** [comment]
From a memory perspective I don't think `i0` makes much sense. `ix` implies that we have `(x-1) bits + 1 signed bit`. `i0` cannot have a signed bit because of its size.

**Comment by winksaville** [comment]
Obviously i0 has no bits, as does u0 and I think the "virtual" value assigined to the type is by convention. So I don't see the need for an explicit sign bit.

Also, the notion of a negative number of bits doesn't seem that different from having no bits.

I'm not trying to be flip, my suggestion is based on the goal that all Types that are members of TypeId.Int should support all operations of every other member. If that goal is not possible and a rational reason can't be found for the deviation from what other members can do, then that type should be considered for removal.

For example, taking the address or determining a structs fields offest in some types is possible while others it is not. I would suggest that if a type has no address or offset then it should be considered for removal.

On the other hand I find it odd that the current lack of support for addresses and offsets for u0. From my perspective u0 always has an address or offset whenever a non-zero type does. And I see no reason i0 is any different.

**Comment by thejoshwolfe** [comment]
JavaScript BigInt allows the equivalent of `i0` which always has the value `0`. https://tc39.github.io/proposal-bigint/#sec-bigint.asintn (`bits` is `0`, `bigint` is any integer, `mod` is 0, step 4 takes the "otherwise" branch.)

I believe the status quo `i1` works reasonably, albeit counterintuitively, where the only values are `0` and `-1`. (Anecdotally, this is how [Booleans work in Visual Basic](https://stackoverflow.com/questions/3621037/casting-a-boolean-to-an-integer-returns-1-for-true); a `Boolean` is effectively an `i1`.)

If we look at the sequence of maximum and minimum values for signed integers as the number of bits approaches 0, we see:

```
 n | min | max
i5 | -16 |  15
i4 |  -8 |   7
i3 |  -4 |   3
i2 |  -2 |   1
i1 |  -1 |   0
i0 |   ? |   ?
```

The pattern here is:

* `min(n) = -2**(n - 1)`. `min(0) = -2**(0 - 1) = -0.5`
* `max(n) = 2**(n - 1) - 1`. `max(0) = 2**(0 - 1) - 1 = -0.5`

Now that I've worked through this on my own, I see that this is exactly what @winksaville reported above.

Nowhere else in the integer types do we see non-integer limits. For unsigned integers this is the formula:

* `min(n) = 0`. `min(0) = 0`
* `max(n) = 2**n - 1`. `max(0) = 2**0 - 1 = 0`

That clearly suggests that `u0` can be equal to `0`. But `i0` being equal to `-0.5` seems very strange.

I do not agree that it makes sense to convert `-0.5` into the integer `0`. Mathematically, there's no compelling reason to choose `0` instead of `-1`. Wikipedia provides [6 different ways](https://en.wikipedia.org/wiki/Rounding#Rounding_to_the_nearest_integer) to deal with rounding half integers to the nearest integer with various arguments for each way. Additionally, you can consider flooring to `-1`, truncating to `0`, or any number of other methods for converting non-integers to integers. My point here is that it's not so obvious that the language should bake the assumption into its semantics.

`i0` is too weird, and doesn't make sense.

**Comment by winksaville** [comment]
I disagree that i0 weird enough to remove at this stage, If there is no reasonable value for i0 then it should be removed. But, I'd say 0 is a reasonable value. If we can agree on the "properties" of TypeId.Int and we can have all iX's and uX's adhere to these properties then that would be best. The consistency of allowing X to range from 0..N where N is now 128 would be preferrable to exceptions for any particular value of X. IMHO.

**Comment by wirelyre** [comment]
Perhaps some fresh eyes can help untangle this.

The only material difference between `i𝑛` and `u𝑛` seems to be comparison. You can rewrite "less than" for signed integers in terms of `@bitCast`ed unsigned integers `T`:

```
if ((a <= @maxValue(T) / 2) and (b <= @maxValue(T) / 2)
  or (a > @maxValue(T) / 2) and (b > @maxValue(T) / 2))
    a < b
else
    a > @maxValue(T) / 2
```

In the case of `u0`, this reduces to `a < b` since the first condition is always true. (Actually, it reduces to `false` since the complete type is `u0 = {0}`, so the only relation is `0 < 0 == false`.)

No need for negative bits or non-integer ranges. They arise from the assumption that the range of signed integers is evenly partitioned into negative and non-negative numbers. This is obviously false if there is only one integer value.

Under this logical scheme, `u0` and `i0` are treated as genuine integer types with a single value `0`, rather than fancy aliases of the unit type.

**Comment by scurest** [comment]
The minimum/maximum makes sense if you phrase the upper bound exclusively (Dijkstra [argues](https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html) this is the preferred way to phrase an interval): `in` is the set { x in Z | -2^(n-1) <= x < 2^(n-1) }, so `i0` is { x in Z | -1/2 <= x < 1/2 } = { 0 }.

Also `@IntType(s, n)` is the image of `@IntType(s, n+1)` under the map of integers x ↦ trunc(x/2) which again suggests `i0` should be {0}. (It also suggests `u0` should be {0}, which it is.)

IMO it also makes sense to have it just for symmetry with `u0`.

**Comment by mnemnion** [comment]
A signed integer is negative if the highest bit is one. If the highest bit is zero, and any other bit is one, it's positive.  Otherwise it is zero.

`i0` is zero. QED.

**Comment by andrewrk** [comment]
https://codeberg.org/ziglang/zig/commit/5d55999d20c4d856ad9e0bd35a939e4aa3be3d0d
